File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:


int f(vector<int>&value)
{
    int n=value.size();
    if (n == 0) return 0;        
    if (n == 1) return value[0];
    vector<int>dp(n,0);

    dp[0]=value[0];
    for(int i=1;i<n;i++)
    {
        int pick=value[i];
        if(i>1)
        pick+=dp[i-2];
        int notpick=0+dp[i-1];
        dp[i]=max(pick,notpick);
    }
    return dp[n-1];
}
    int rob(vector<int>& nums) {
        vector<int>temp1,temp2;
        int n =nums.size();
        if(n==1)return nums[0];
        for(int i=0;i<n;i++)
        {
            if(i!=0)temp1.push_back(nums[i]);
            if(i!=n-1)temp2.push_back(nums[i]);
        }
        return max(f(temp1),f(temp2));
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int f(int[]nums, int s,int end,int[]dp)
    {
        if(s>end)return 0;

        if(dp[s]!=-1)return dp[s];

        int take=nums[s]+f(nums,s+2,end,dp);
        int notake=f(nums,s+1,end,dp);

        return dp[s]=Math.max(take,notake);

    }
    public int rob(int[] nums) {
        int n=nums.length;
        
       

        if(n==1)return nums[0];

        int[] dp1 = new int[n];
        Arrays.fill(dp1, -1);

        int case1=f(nums,0,n-2,dp1);


        int[] dp2 = new int[n];
        Arrays.fill(dp2, -1);
        int case2=f(nums,1,n-1,dp2);

        return Math.max(case1,case2);
    }
}